src/renderer/canvas/edge/base.ts

```diff
@@ -145,34 +147,38 @@ const setupShadow = <N extends INodeBase, E extends IEdgeBase>(
   context: CanvasRenderingContext2D,
   edge: IEdge<N, E>,
 ) => {
-  if (edge.style.shadowColor) {
-    context.shadowColor = edge.style.shadowColor.toString();
+  const edgeStyle = edge.getStyle();
+
+  if (edgeStyle.shadowColor) {
+    context.shadowColor = (edgeStyle.shadowColor as string | Color).toString();
   }
-  if (edge.style.shadowSize) {
-    context.shadowBlur = edge.style.shadowSize;
+  if (edgeStyle.shadowSize) {
+    context.shadowBlur = edgeStyle.shadowSize as number;
   }
-  if (edge.style.shadowOffsetX) {
-    context.shadowOffsetX = edge.style.shadowOffsetX;
+  if (edgeStyle.shadowOffsetX) {
+    context.shadowOffsetX = edgeStyle.shadowOffsetX as number;
   }
-  if (edge.style.shadowOffsetY) {
-    context.shadowOffsetY = edge.style.shadowOffsetY;
+  if (edgeStyle.shadowOffsetY) {
+    context.shadowOffsetY = edgeStyle.shadowOffsetY as number;
```

Comment:
Why these `as`? I checked in VSCode - aren't those properties already numbers?

If you need strictly numbers, you can use nullish coalescing to get rid of `undefined` -> `edgeStyle.shadowSize = edgeStyle.shadowSize ?? 0`. Maybe then these if conditions aren't necessary then. Not sure if we want to set these values to 0 if they aren't defined.

Reply:
Good catch for the `as number`. Btw conditions are here to dismiss any unnecessary setup of context variables because even setting `context.shadowOffsetY = 0` will do something. These conditions were here just to setup context only if it is any non-zero number.